src/game/WorldSession.cpp

```diff
@@ -102,17 +99,16 @@ WorldSession::~WorldSession()
         LogoutPlayer(true);
 
     /// - If have unclosed socket, close it
-    if (m_Socket)
-    {
-        m_Socket->CloseSocket();
-        m_Socket->RemoveReference();
-        m_Socket = nullptr;
-    }
+    if (!m_Socket->IsClosed())
+        m_Socket->Close();
 
     ///- empty incoming packet queue
-    WorldPacket* packet = nullptr;
-    while (_recvQueue.next(packet))
+    while (!m_packetQueue.empty())
```

Comment:
Okay, I misunderstood you before.  I thought you were saying that there could be a double use of the queue.  I think that that was possible too, and that is what the commit I linked above should address.

If I understand you correctly now, the potential problem scenario is if World::UpdateSessions deletes the WorldSession at the same time that WorldSocket::ProcessIncomingData is executing, but only after the IsClosed() check.  Yes, that could be a problem.  Was it not a problem with the ACE networking code?  The only way I could see it not having been a problem before is if WorldSocket::ProcessIncomingData and World::UpdateSessions were guaranteed to be called from the same thread.